sparse/_coo/common.py

```diff
@@ -1215,6 +1215,127 @@ def ones_like(a, dtype=None):
     return ones(a.shape, dtype=(a.dtype if dtype is None else dtype))
 
 
+def diagonal(a, offset=0, axis1=0, axis2=1):
+    """
+    Extract diagonal from a COO array. The equivalent of :obj:`numpy.diagonal`.
+
+    Parameters
+    ----------
+    a: COO
+        The array to perform the operation on.
+    offset: int, optional
+        Offset of the diagonal from the main diagonal. Defaults to main diagonal (0).
+    axis1: int, optional
+        First axis from which the diagonals should be taken.  
+        Defaults to first axis (0).
+    axis2 : int, optional
+        Second axis from which the diagonals should be taken.  
+        Defaults to second axis (1).
+
+    Examples
+    --------
+    >>> import sparse
+    >>> x = sparse.as_coo(np.arange(9).reshape(3,3))
+    >>> sparse.diagonal(x).todense()
+    array([0, 4, 8])
+    >>> sparse.diagonal(x,offset=1).todense()
+    array([1, 5])
+
+    >>> x = sparse.as_coo(np.arange(12).reshape((2,3,2)))
+    >>> x_diag = sparse.diagonal(x, axis1=0, axis2=2)
+    >>> x_diag.shape
+    (3, 2)
+    >>> x_diag.todense()
+    array([[ 0,  7],
+           [ 2,  9],
+           [ 4, 11]])
+
+    Returns
+    -------
+    out: COO
+        The result of the operation.
+
+    Raises
+    ------
+    ValueError
+        If a.shape[axis1] != a.shape[axis2]
+
+    See Also
+    --------
+    :obj:`numpy.diagonal`: NumPy equivalent function
+    """
+    from .core import COO
+
+    if a.shape[axis1] != a.shape[axis2]:
+        raise ValueError("a.shape[axis1] != a.shape[axis2]")
+
+    diag_axes = [
+        axis for axis in range(len(a.shape)) if axis != axis1 and axis != axis2
+    ] + [axis1]
+    diag_shape = [a.shape[axis] for axis in diag_axes]
+    diag_shape[-1] -= abs(offset)
+
+    diag_idx = _diagonal_idx(a.coords, axis1, axis2, offset)
+
+    diag_coords = [a.coords[axis][diag_idx] for axis in diag_axes]
+    diag_data = a.data[diag_idx]
+
+    return COO(diag_coords, diag_data, diag_shape)
+
+
+def diagonalize(a, axis=0):
+    """
+    Diagonalize a COO array. The new dimension is appended at the end.
+
+    **Warning**: :obj:`diagonalize` is not :obj:`numpy` compatible as there is no direct :obj:`numpy` equivalent. The API may change in the future.
```

Comment:
This should be in a [`..warning::` directive](http://docutils.sourceforge.net/docs/ref/rst/directives.html#warning).